addons/payment_authorize/models/payment.py

```diff
@@ -56,12 +67,23 @@ def _authorize_generate_hashing(self, values):
             values['x_amount'],
             values['x_currency_code']])
 
-        # [BACKWARD COMPATIBILITY] Check that the merchant did update his transaction
-        # key to signature key (end of MD5 support from Authorize.net)
+        # [BACKWARD COMPATIBILITY, 2nd edition]
         # The signature key is now '128-character hexadecimal format', while the
         # transaction key was only 16-character.
-        if len(values['x_trans_key']) == 128:
+        # One of 2 things should have happened:
+        # 1/ the Transaction Key has been replaced with the Signature Key value (patch from March 2019)
+        #       => Use that to sign, but server-to-server won't work since it uses transaction key
+        #          as its credentials
+        # 2/ the Signature key is a new field (patch from July 2019)
+        #       => Use that field for the signature
+
+        # FORWARD-PORT NOTE: be careful, hexadecimal decoding in python 3 is done by using bytes.fromhex(str)
+        # (P2) self.authorize_signature_key.decode("hex") ==> (P3) bytes.fromhex(self.authorize_signature_key)
+        if len(values['x_trans_key']) == 128 and not self.authorize_signature_key:
```

Comment:
> Since both flows can coexist on the same configuration

Does it means that both flow (s2s and form redirect) will go through this method? The s2s still need to `MD5` encrypt the `signature key` while the form redirect need to `SHA` encrypt the `signature key`, right?

But if the user has configured both keys, it will never go through the `else` condition, unless I am missing something

Reply:
No, s2s does not use any hashing of any field - it just sends the api login and transaction key as headers for credentials. Since it communicates directly with authorize.net servers over https, there is no need for hashing. The hashing is only useful for form flow.

If the user has configured both keys, the second statement will be executed (since the transaction key is not 128 char long in that case). The first will only be executed once if the transaction key is 128 chars long (that way fields keep their semantic use and the 128char key is set on the signature field).

We could honestly remove the last else statement since md5 is dead anyway; but i'd rather keep these for master.

Reply:
Thanks for the clarification :+1: